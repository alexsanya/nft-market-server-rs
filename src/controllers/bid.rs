use tokio::join;
use crate::{models::bid::Bid, prelude::Result, repositories::bids::{get_all, save_bid}, services::onchain_provider::{ensure_bidder_has_tokens, ensure_erc20_allowance, ensure_owner_has_nft, esnure_erc712_allowance}};

pub async fn create_bid(bid: &Bid) -> Result<()> {
    bid.verify_signature()?;
    let (nft_ownership_check, nft_allowance_check, tokens_check, allowance_check)  = join!(
        ensure_owner_has_nft(&bid.listing.owner, &bid.listing.nft_contract, &bid.listing.token_id),
        esnure_erc712_allowance(&bid.listing.owner, &bid.listing.nft_contract),
        ensure_bidder_has_tokens(&bid.bidder, &bid.token_address, &bid.value),
        ensure_erc20_allowance(&bid.bidder, &bid.token_address, &bid.value)
    );
    nft_ownership_check?;
    nft_allowance_check?;
    tokens_check?;
    allowance_check?;
    save_bid(bid).await?;
    Ok(())
}

pub fn get_bids() -> Result<Vec<Bid>> {
    get_all()
}
