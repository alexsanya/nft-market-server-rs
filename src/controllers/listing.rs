use crate::{models::listing::Listing, repositories::listing::{get_all, save_listing}};


pub async fn create_listing(listing: &Listing) -> Result<(), ()> {
    save_listing(listing).await?;
    Ok(())
}

pub fn get_listings() -> Vec<Listing> {
    get_all().unwrap()
}