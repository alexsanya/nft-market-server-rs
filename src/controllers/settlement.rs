use tokio::join;
use crate::{models::settlement::Settlement, prelude::Result, repositories::settlement::{get_all, save_settlement}, services::onchain_provider::{ensure_bidder_has_tokens, ensure_erc20_allowance, ensure_owner_has_nft, esnure_erc712_allowance}};

pub async fn create_settlement(settlement: &Settlement) -> Result<()> {
    settlement.verify_signature()?;
    let (nft_ownership_check, nft_allowance_check, tokens_check, allowance_check)  = join!(
        ensure_owner_has_nft(&settlement.bid.listing.owner, &settlement.bid.listing.nft_contract, &settlement.bid.listing.token_id),
        esnure_erc712_allowance(&settlement.bid.listing.owner, &settlement.bid.listing.nft_contract),
        ensure_bidder_has_tokens(&settlement.bid.bidder, &settlement.bid.token_address, &settlement.bid.value),
        ensure_erc20_allowance(&settlement.bid.bidder, &settlement.bid.token_address, &settlement.bid.value)
    );
    nft_ownership_check?;
    nft_allowance_check?;
    tokens_check?;
    allowance_check?;
    save_settlement(settlement).await?;
    Ok(())
}

pub fn get_settlements() -> Result<Vec<Settlement>> {
    get_all()
}
