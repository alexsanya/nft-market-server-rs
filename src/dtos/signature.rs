
use serde::{Serialize, Deserialize};

use crate::{models::signature::Signature, utils::patterns::Patterns};

#[derive(Debug, Serialize, Deserialize)]
pub struct SignatureDTO {
    pub v: u64,
    pub r: String,
    pub s: String
}

#[derive(Debug, Clone)]
pub enum ParsingError {
    RisInvalid,
    SisInvalid
}

impl TryInto<Signature> for SignatureDTO {
    type Error = ParsingError;

    fn try_into(self) -> Result<Signature, Self::Error> {
        let patterns = Patterns::new();
        Ok(
            Signature{
                v: self.v,
                r: patterns.test_bytes32(&self.r).map_err(|_| ParsingError::RisInvalid)?.to_owned(),
                s: patterns.test_bytes32(&self.s).map_err(|_| ParsingError::SisInvalid)?.to_owned()
            }
        )
    }
}