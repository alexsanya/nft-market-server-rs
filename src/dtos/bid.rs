use serde::{Deserialize, Serialize};
use crate::models::bid::Bid;
use crate::utils::patterns::Patterns;
use super::listing::ListingDTO;
use super::signature::SignatureDTO;
use super::listing::ParsingError as ParsingListingError;
use super::signature::ParsingError as ParsingSignatureError;

#[derive(Debug, strum_macros::AsRefStr)]
pub enum ParsingError {
    Bidder,
    TokenAddress,
    ValidUntil,
    Value,
    Listing(ParsingListingError),
    Signature(ParsingSignatureError)
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BidDTO {
    pub bidder: String,
    pub listing: ListingDTO,
    pub token_address: String,
    pub valid_until: String,
    pub value: String,
    pub signature: SignatureDTO
}

impl TryInto<Bid> for BidDTO {
    type Error = ParsingError;

    fn try_into(self) -> Result<Bid, Self::Error> {
        let patterns = Patterns::new();
        let bid = Bid {
            bidder: patterns.test_address(&self.bidder).map_err(|_| ParsingError::Bidder)?.to_owned(),
            listing: self.listing.try_into().map_err(ParsingError::Listing)?,
            token_address: patterns.test_address(&self.token_address).map_err(|_| ParsingError::TokenAddress)?.to_owned(),
            valid_until: self.valid_until.parse().map_err(|_| ParsingError::ValidUntil)?,
            value: self.value.parse().map_err(|_| ParsingError::Value)?,
            siganture: self.signature.try_into().map_err(ParsingError::Signature)?
        };
        Ok(bid)
    }
}

#[cfg(test)]
mod test {
    use serde_json::json;
    use insta::assert_debug_snapshot;

    use super::*;

    #[test]
    fn convert() {
        let raw_listing = json!({
            "owner": "0x3897326cEda92B3da2c27a224D6fDCFefCaCf57A",
            "chain_id": "11155111",
            "min_price_cents": "150000",
            "nft_contract": "0xf44b599a0aB6b8cb14E992994BEC0dc59dF883B2",
            "token_id": "1",
            "nonce": "0",
            "signature": {
                "v": 28,
                "r": "0x5ef4620f4b296763ff15209456d75e868f149a8d1c6821f1ff11fab70bca0ee0",
                "s": "0x337ddcb26ea919a2bf5ad6e1d49bd6951a27d1d2e940a5543a70eabc5dbe237e"
            }
        });

        let raw_bid = json!({
            "bidder": "0xE98D94496aB9084f597a69978b593EBf83147335",
            "listing": raw_listing,
            "token_address": "0xc29f6F8D639eF187DcFEfeFBaD989cF2C941a23A",
            "valid_until": "1735504160",
            "value": "250",
            "signature": {
                "v": 28,
                "r": "0x1469ac6f9636c24d2d8c3fb2cbef73708876e15f23f23b1d33863939c905a21c",
                "s": "0x7d9a7ea039465c928311bcb737b23153232028038beadba2a667aa720f17602b"
            }
        });

        let bid_dto: BidDTO = serde_json::from_value(raw_bid).expect("Cannot convert jsonBid to BidDTO");
        let bid: Bid = bid_dto.try_into().expect("Cannot convert BidBTo to bid");
        assert_debug_snapshot!(bid);
    }
}