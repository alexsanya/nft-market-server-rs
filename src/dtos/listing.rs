use serde::{Serialize, Deserialize};
use crate::{models::listing::Listing, utils::patterns::Patterns};
use super::signature::{self, SignatureDTO};

#[derive(Debug)]
pub enum ParsingError {
    Owner,
    ChainId,
    NftContract,
    TokenId,
    Nonce,
    Signature(signature::ParsingError)
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ListingDTO {
    pub owner: String,
    pub chain_id: String,
    pub min_price_cents: u16,
    pub nft_contract: String,
    pub token_id: String,
    pub nonce: String,
    pub signature: SignatureDTO
}

impl TryInto<Listing> for ListingDTO {
    type Error = ParsingError;

    fn try_into(self) -> Result<Listing, Self::Error> {
        let patterns = Patterns::new();
        let listing = Listing{
            owner: patterns.test_address(&self.owner).map_err(|_| ParsingError::Owner)?.to_owned(),
            chain_id: self.chain_id.parse().map_err(|_| ParsingError::ChainId)?,
            min_price_cents: self.min_price_cents,
            nft_contract: patterns.test_address(&self.nft_contract).map_err(|_| ParsingError::NftContract)?.to_owned(),
            token_id: self.token_id.parse().map_err(|_| ParsingError::TokenId)?,
            nonce: self.nonce.parse().map_err(|_| ParsingError::Nonce)?,
            signature: self.signature.try_into().map_err(ParsingError::Signature)?
        };
        Ok(listing)
    }
}