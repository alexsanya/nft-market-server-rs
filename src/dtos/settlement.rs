use serde::{Deserialize, Serialize};
use crate::models::settlement::Settlement;
use crate::dtos::listing::ParsingError as ListingParsingError;
use crate::dtos::bid::ParsingError as BidParsingError;
use crate::dtos::signature::ParsingError as SigParsingError;

use super::{bid::BidDTO, listing::ListingDTO, signature::SignatureDTO};

#[derive(Debug, Serialize, Deserialize)]
pub struct SettlementDTO {
    pub listing: ListingDTO,
    pub bid: BidDTO,
    pub signature: SignatureDTO
}


#[derive(Debug, Clone, strum_macros::AsRefStr)]
pub enum ParsingError {
    Listing(ListingParsingError),
    Bid(BidParsingError),
    Signature(SigParsingError)
}

impl TryInto<Settlement> for SettlementDTO {
    type Error = ParsingError;

    fn try_into(self) -> Result<Settlement, Self::Error> {
        Ok(Settlement {
            listing: self.listing.try_into().map_err(ParsingError::Listing)?,
            bid: self.bid.try_into().map_err(ParsingError::Bid)?,
            signature: self.signature.try_into().map_err(ParsingError::Signature)?
        })
    }
}

#[cfg(test)]
mod test {
    use insta::assert_debug_snapshot;

    use crate::models::samples::test::get_raw_settlement;

    use super::*;

    #[test]
    fn convert() {
        let raw_settlement = get_raw_settlement();
        let settlement_dto: SettlementDTO = serde_json::from_value(raw_settlement).expect("Cannot convert json to SettlementDTO");
        let settlement: Settlement = settlement_dto.try_into().expect("Cannot convert SettlementDTO to settlement");
        assert_debug_snapshot!(settlement);
    }
}