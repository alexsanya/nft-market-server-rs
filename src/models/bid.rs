use ethers::types::{Address, Signature, U256};
use num_bigint::BigInt;
use std::str::FromStr;
use serde::{Deserialize, Serialize};
use super::signature::Signature as SigString;
use super::bid_eip712::Bid as BidEIP712;
use super::listing_eip712::Listing as ListingEIP712;
use ethers::types::transaction::eip712::Eip712;
use crate::error::{Entity, Error};
use crate::prelude::Result as MyResult;
use super::listing::Listing;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Bid {
    pub bidder: String,
    pub listing: Listing,
    pub token_address: String,
    pub valid_until: BigInt,
    pub value: BigInt,
    pub signature: SigString
}

#[derive(Debug)]
pub enum Eip712Error {
    TokenAddress,
    ValidUntil,
    Value,
    Listing
}

impl TryInto<BidEIP712> for Bid {
    type Error = Eip712Error;
    fn try_into(self) -> Result<BidEIP712, Self::Error> {
        let listing_eip712: ListingEIP712 = self.listing.try_into().map_err(|_| Eip712Error::Listing)?;
        let hash = listing_eip712.encode_eip712().map_err(|_| Eip712Error::Listing)?;
        let bid_eip712 = BidEIP712 {
            token_contract: Address::from_str(&self.token_address).map_err(|_| Eip712Error::TokenAddress)?,
            value: U256::from_dec_str(&self.value.to_string()).map_err(|_| Eip712Error::Value)?,
            valid_until: U256::from_dec_str(&self.valid_until.to_string()).map_err(|_| Eip712Error::ValidUntil)?,
            listing_hash: hash.into()
        };
        Ok(bid_eip712)
    }
}

impl Bid {
    pub fn verify_signature(&self) -> MyResult<()> {
        self.listing.verify_signature()?;
        let signature: Signature = self.signature.clone().try_into().map_err(|_| Error::InvalidSignature(Entity::Bid))?;
        let bid_eip712: BidEIP712 = self.clone().try_into().map_err(|_| Error::InvalidSignature(Entity::Bid))?;
        let recover_result = signature.recover_typed_data(&bid_eip712);
        let bidder_result = Address::from_str(&self.bidder);
        match (recover_result, bidder_result) {
            (Ok(recovered), Ok(bidder)) if recovered == bidder => Ok(()),
            _ => Err(Error::InvalidSignature(Entity::Bid))
        }
    }
}


#[cfg(test)]
mod tests {
    use std::str::FromStr;
    use ethers::types::Signature;
    use ethers::types::Address;
    use crate::models::samples::test::get_bid;
    use super::*;

    #[test]
    fn check_hash() {
        let bid = get_bid();
        let signature: Signature = bid.signature.clone().try_into().expect("Cannot convert signature");
        let bidder = bid.bidder.clone();
        let bid_eip712: BidEIP712 = bid.try_into().expect("Cannot convert Bid into BidEIP712");
        let address = signature.recover_typed_data(&bid_eip712).expect("Cannot recover typed data");
        println!("Recovered address: {:?}", address);
        assert_eq!(address, Address::from_str(&bidder).unwrap());

    }

    #[test]
    fn check_signature_ok() {
        let bid = get_bid();
        bid.verify_signature().expect("Error");
    }

    #[test]
    fn check_listing_sig_incorrect() {
        let bid = get_bid();
        let bid_incorrect_sig = Bid {
            listing: Listing {
                signature: SigString {
                    v: 29,
                    ..bid.signature.clone()
                },
                ..bid.listing
            },
            ..bid
        };
        assert!(matches!(bid_incorrect_sig.verify_signature(), Err(Error::InvalidSignature(Entity::Listing))));
    }

    #[test]
    fn check_bid_sig_incorrect() {
        let bid = get_bid();
        let bid_incorrect_sig = Bid {
            signature: SigString {
                v: 29,
                ..bid.signature
            },
            ..bid
        };
        assert!(matches!(bid_incorrect_sig.verify_signature(), Err(Error::InvalidSignature(Entity::Bid))));
    }
}