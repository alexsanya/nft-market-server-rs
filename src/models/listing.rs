use num_bigint::BigInt;
use crate::utils::hash::Hashable;
use ethers::utils::keccak256;
use ethers::abi::AbiEncode;
use tracing::debug;
use hex;

use super::signature::Signature;
use serde::{Serialize, Deserialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Listing {
    pub owner: String,
    pub chain_id: BigInt,
    pub min_price_cents: u16,
    pub nft_contract: String,
    pub token_id: BigInt,
    pub nonce: BigInt,
    pub signature: Signature
}

impl Hashable for Listing {
    fn hash(&self, domain_separator: String) -> String {
        let encoded_data = (
            keccak256(b"Listing(address nftContract,uint256 tokenId,uint256 minPriceCents,uint256 nonce)"),
            self.nft_contract.parse::<ethers::types::Address>().unwrap(),
            self.token_id.to_string().parse::<ethers::types::U256>().unwrap(),
            hex::encode(self.min_price_cents.to_be_bytes()).parse::<ethers::types::U256>().unwrap(),
            self.nonce.to_string().parse::<ethers::types::U256>().unwrap()
        ).encode();

        print!("hello hash: {}", hex::encode(keccak256(b"hello")));
        print!("encoded_data: {}", hex::encode(&encoded_data));
        print!("encoded_data hash: {:?}", hex::encode(keccak256(&encoded_data[..])));

        //let hex_raw = (
        //    ethers::types::Bytes::from(vec![0x19, 0x01]),
        //    domain_separator.parse::<ethers::types::Bytes>().unwrap(),
        //    ethers::types::Bytes::from(keccak256(encoded_data))
        //).encode();

        let prefix = vec![0x19, 0x01];
        let result = [&prefix[..], &hex::decode(domain_separator).unwrap()[..], &keccak256(&encoded_data)[..]].concat();

        print!("hex_raw: {}", hex::encode(&result));
        hex::encode(keccak256(&result))
    }
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use crate::dtos::listing::ListingDTO;

    use super::*;

    #[test]
    fn check_hash() {
        let raw_listing = json!({
            "owner": "0x3897326cEda92B3da2c27a224D6fDCFefCaCf57A",
            "chain_id": "11155111",
            "min_price_cents": 137,
            "nft_contract": "0xf44b599a0aB6b8cb14E992994BEC0dc59dF883B2",
            "token_id": "1",
            "nonce": "0",
            "signature": {
                "v": 28,
                "r": "0x5ef4620f4b296763ff15209456d75e868f149a8d1c6821f1ff11fab70bca0ee0",
                "s": "0x337ddcb26ea919a2bf5ad6e1d49bd6951a27d1d2e940a5543a70eabc5dbe237e"
            }
        });

        let DOMAIN_SEPARATOR = "47720df067349d6e15f380966605c432d4f59c1ad2d55501d9d8ea139c7244d9".to_owned();

        let listing_dto: ListingDTO = serde_json::from_value(raw_listing).unwrap();

        let listing: Listing = listing_dto.try_into().unwrap();

        let hash = listing.hash(DOMAIN_SEPARATOR);

        assert_eq!(hash, "07e440212245f786b1ab066a629db7ef7b9e6f98b1ced217ed260b3d1e1a1fa3");

    }
}