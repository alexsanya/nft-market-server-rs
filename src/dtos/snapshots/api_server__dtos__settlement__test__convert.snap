---
source: src/dtos/settlement.rs
expression: settlement
snapshot_kind: text
---
Settlement {
    listing: Listing {
        owner: "0x3897326cEda92B3da2c27a224D6fDCFefCaCf57A",
        chain_id: 11155111,
        min_price_cents: 150000,
        nft_contract: "0xf44b599a0aB6b8cb14E992994BEC0dc59dF883B2",
        token_id: 1,
        nonce: 0,
        signature: Signature {
            v: 28,
            r: "0x5ef4620f4b296763ff15209456d75e868f149a8d1c6821f1ff11fab70bca0ee0",
            s: "0x337ddcb26ea919a2bf5ad6e1d49bd6951a27d1d2e940a5543a70eabc5dbe237e",
        },
    },
    bid: Bid {
        bidder: "0xE98D94496aB9084f597a69978b593EBf83147335",
        listing: Listing {
            owner: "0x3897326cEda92B3da2c27a224D6fDCFefCaCf57A",
            chain_id: 11155111,
            min_price_cents: 150000,
            nft_contract: "0xf44b599a0aB6b8cb14E992994BEC0dc59dF883B2",
            token_id: 1,
            nonce: 0,
            signature: Signature {
                v: 28,
                r: "0x5ef4620f4b296763ff15209456d75e868f149a8d1c6821f1ff11fab70bca0ee0",
                s: "0x337ddcb26ea919a2bf5ad6e1d49bd6951a27d1d2e940a5543a70eabc5dbe237e",
            },
        },
        token_address: "0xc29f6F8D639eF187DcFEfeFBaD989cF2C941a23A",
        valid_until: 1735504160,
        value: 250,
        signature: Signature {
            v: 28,
            r: "0x1469ac6f9636c24d2d8c3fb2cbef73708876e15f23f23b1d33863939c905a21c",
            s: "0x7d9a7ea039465c928311bcb737b23153232028038beadba2a667aa720f17602b",
        },
    },
    signature: Signature {
        v: 27,
        r: "0xc21f88f00f01849ecbe4bcb75bd8f6cc2ac1f3507498e385b78df7db5f5ae334",
        s: "0x6adc46861e9888247b6b4f55cd7eb73449d835c94a0c3fd5e2df1b8cb6f77c4c",
    },
}
